crates/bevy_math/src/primitives/dim2.rs

```diff
@@ -28,6 +28,30 @@ impl Direction2d {
         )
     }
 
+    /// Create a direction from a finite, nonzero [`Vec2`], also returning its original length.
+    ///
+    /// Returns [`Err(InvalidDirectionError)`](InvalidDirectionError) if the length
+    /// of the given vector is zero (or very close to zero), infinite, or `NaN`.
+    pub fn new_and_length(value: Vec2) -> Result<(Self, f32), InvalidDirectionError> {
+        let len = value.length();
+        let direction = (len.is_finite() && len > 0.0).then_some(value / len);
+
+        direction.map(|dir| (Self(dir), len)).map_or_else(
+            || {
+                if len.is_nan() {
```

Comment:
This block is the same between 2d and 3d. Might be worth moving to a method on `InvalidDirectionError`.

Reply:
I added an `InvalidDirectionError::from_length` method for this.

I also realized that I can use `new_and_length` for the `new` constructors, which cleans things up quite nicely. I believe the performance should be equivalent, since the previously used `try_normalized` had to get the length anyway and do the same computations.